filewatcher/filewatcher.go

```diff
@@ -107,26 +114,64 @@ func (r *Result) watcherLoop(
 			default:
 				// Ignore uninterested events.
 			case fsnotify.Create, fsnotify.Write:
-				// Wrap with an anonymous function to make sure that defer works.
-				func() {
-					f, err := limitopen.OpenWithLimit(path, softLimit, hardLimit)
-					if err != nil {
-						logger.Log(context.Background(), "filewatcher: I/O error: "+err.Error())
-						return
-					}
-					defer f.Close()
-					d, err := parser(f)
-					if err != nil {
-						logger.Log(context.Background(), "filewatcher: parser error: "+err.Error())
-					} else {
-						r.data.Store(d)
-					}
-				}()
+				stat, err := os.Stat(r.path)
+				if err != nil {
+					r.logger.Log(context.Background(), fmt.Sprintf(
+						"filewatcher: failed to get mtime for %q: %v",
+						r.path,
+						err,
+					))
+					continue
+				}
+				r.forceReload(stat.ModTime())
 			}
 		}
 	}
 }
 
+func (r *Result) reload() {
+	stat, err := os.Stat(r.path)
+	if err != nil {
+		r.logger.Log(context.Background(), fmt.Sprintf(
+			"filewatcher: failed to get mtime for %q: %v",
+			r.path,
+			err,
+		))
+		return
+	}
+	mtime := stat.ModTime()
+	if d, _ := r.data.Load().(*atomicData); d == nil || d.mtime.Before(mtime) {
+		r.forceReload(mtime)
+	}
+}
+
+func (r *Result) forceReload(mtime time.Time) {
+	f, err := limitopen.OpenWithLimit(r.path, r.softLimit, r.hardLimit)
+	if err != nil {
+		r.logger.Log(context.Background(), "filewatcher: I/O error: "+err.Error())
+		return
+	}
+	defer f.Close()
+	d, err := r.parser(f)
+	if err != nil {
+		r.logger.Log(context.Background(), "filewatcher: parser error: "+err.Error())
+	} else {
+		r.data.Store(&atomicData{
+			data:  d,
+			mtime: mtime,
```

Comment:
I think we probably want to re-grab the mtime from the file rather than using the one that's passed in case the file was updated between when we checked the mtime and when we opened the file?

Reply:
in that scenario we just do an extra `forceReload` and load in the same data, which is not a big deal.